### ISSUE TITLE
S3 Trio64 not emulated well enough for Windows 3.1x Trio64 drivers

### ISSUE THREAD

**Issue by FeralChild64** [dosbox-staging/dosbox-staging#2767]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

current GIT main (0.81.0-alpha) - e410259b43c3bf3fd61eb42d390d43007e5e1c81

### What Operating System are you using?

Linux x86_64

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Since the configuration file claims Staging emulates the S3 Trio64 card, we should IMHO have this issue open so that the problem is not forgotten.

There is at least one Windows 3.1x game - Critical Mass by Sean O'Connir - which exposes a Trio64 emulation problem if used with S3 Trio 32/64 drivers v1.70.04:

![S3_Trio64_1 70 04](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/1e871946-5c6c-45fd-8a90-5adc0fd6844c)

Observe the Alpha Squadron panel window on the right side of the screen - it's misses the ships, they are not drawn. 

Here is how the game should look like - screenshot taken using S3 Vision964 drivers v1.41B5:

![S3_964_1 41B5](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/2ec52d53-8f26-4b96-af41-cb724389c46a)

### Steps to reproduce the behaviour.

1. Install Windows 3.11 for Workgroups
2. Install S3 Trio64 driver v1.70.04, resolution 1024x768 64K SF
3. Start Crtitical Mass game demo, press OK, then BEGIN MISSION
4. Observe the Alpha Squadron window on the right side of the screen

### Download URL of affected game or software

https://archive.org/details/CriticalMass_1020

### Your configuration

```ini
[dosbox]
machine = svga_s3
vmemsize = 4
```

### Provide a Log

```text
Extracted errors from the log (full log gets really long really quickly):

2023-08-17 20:00:00.825 | XGA: Wrote to port 8104 with 3d700de, IO width=4
2023-08-17 20:00:00.825 | XGA: Wrote to port 810c with 3df00e8, IO width=4
2023-08-17 20:00:00.825 | XGA: Unhandled draw command 3
2023-08-17 20:00:00.825 | XGA: Wrote to port 8104 with 3d400e6, IO width=4
2023-08-17 20:00:00.825 | XGA: Wrote to port 810c with 3da00e6, IO width=4
2023-08-17 20:00:00.825 | XGA: Unhandled draw command 3
2023-08-17 20:00:00.825 | XGA: Wrote to port 810c with 3da00ec, IO width=4
2023-08-17 20:00:00.825 | XGA: Unhandled draw command 3
2023-08-17 20:00:00.825 | XGA: Wrote to port 8104 with 3d700df, IO width=4
2023-08-17 20:00:00.825 | XGA: Wrote to port 810c with 3d900e3, IO width=4
2023-08-17 20:00:00.825 | XGA: Unhandled draw command 3
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by Grounded0** [comment]
What's your take on the issue. Should this be closed as a bad driver issue or not?

**Comment by FeralChild64** [comment]
AFAIK this is an emulation issue.

**Comment by rderooy** [comment]
DOSBox-X originally also had various issues with the Windows 3.x drivers, but @joncampbell123 was able to fix them. So this is definitely an emulation issue.

**Comment by Grounded0** [comment]
Since its an emulation issue the workaround for now is to use S3 Vision964 driver v1.41B5 listed in DOSBox Staging Wiki Windows page for anyone reading this or facing similar glitches elsewhere in Windows 3.1x.

**Comment by johnnovak** [comment]
> Since its an emulation issue the workaround for now is to use S3 Vision964 driver v1.41B5 listed in DOSBox Staging Wiki Windows page for anyone reading this or facing similar issues elsewhere in Windows 3.1x.

Yeah, I'd say while fixing the S3 Trio64 compatibility issue would be good, from a practical viewpoint if we have a single driver working perfectly with Win 3.x that's all we need.

**Comment by Grounded0** [comment]
For anyone interested why there's this difference between the driver robustness it seems that the S3 Vision964 v1.41B5 driver only partially implements the XGA routines supported by the graphics core. Later drivers implement more XGA routines that our current S3 Trio emulation doesn't have implemented yet.

**Comment by joncampbell123** [comment]
Not only do newer cards add XGA commands but some bitfield widths change slightly as well.

S3 virge completely changes most of the XGA registers and commands.

You may borrow from DOSBox-X if it helps S3 emulation.

**Comment by joncampbell123** [comment]
My collection of S3 datasheets might also help here. http://hackipedia.org/browse.cgi/Computer/Platform/PC%2c%20IBM%20compatible/Video/VGA/SVGA/S3%20Graphics%2c%20Ltd

**Comment by johnnovak** [comment]
Thanks @joncampbell123. Do you know, does XGA support cause any user-observable difference or improvement, or is it just a technical curiosity?

**Comment by Grounded0** [comment]
Accelerated routines 4x to 8x faster even under emulation:

<img width="840" alt="Screenshot 2023-12-16 at 15 15 44" src="https://github.com/dosbox-staging/dosbox-staging/assets/48192273/4437c0ba-312d-48fd-b31a-56077cf3c71e">

I understood GDI performance is still sluggish under Windows 98 SE so implementing all the routines probably would improve drawing performance under Windows 98 SE at least.

**Comment by johnnovak** [comment]
Whoa, that's a lot faster, indeed @Grounded0. Although, doing things correctly might slow it down a bit perhaps 😄

**Comment by interloper98** [comment]
I'm having a hard time getting those green ships to show up in the squadron squares, even under DOSBox-X. I'm using:
- Latest DOSBox-X Visual Studio build from CI ([link](https://github.com/joncampbell123/dosbox-x/actions/runs/9537899983))
- Clean installation of Windows 3.11 WFW (under DOSBox-X) using `vmemsize = 4` and `cycles = 22000` 
- Installed S3 Trio64V+ Win 3.1 English v. 1.70.04 [from vogons library](http://vogonsdrivers.com/getfile.php?fileid=275&menustate=18,1) 1024x768 65K Small fonts
- Installed Critical Mass 1020 version [from archive.org](https://archive.org/details/CriticalMass_1020)

![2024-06-16 15_20_40-cmd (Admin)](https://github.com/dosbox-staging/dosbox-staging/assets/165201780/6fe0bad1-6213-41ba-afc9-e9e13d1fc12a)

@rderooy and @FeralChild64 - have either of you had success with this game using the 1.70 drivers on DOSBox-X, PCem, or 86Box? 

Hopefully this is just user-error on my part!

**Comment by johnnovak** [comment]
Could be a buggy driver & game combo too. I could test in on hardware, but then I have like 4-5 S3 cards with different chip revisions, and it's quite plausible that some drivers are not 100% working on some chipset revisions.

All in all, it's a very minor issue, and if we have a single driver that runs all the games, we're good.

**Comment by interloper98** [comment]
> All in all, it's a very minor issue, and if we have a single driver that runs all the games, we're good.

100% ; I was thinking (hoping?) this game was like an XGA tester to show where we've got functional holes; but after getting it running in DOSBox-X, maybe this isn't the case after all! DOSBox-X doesn't even log any XGA issues on the console, either - so indeed, might just be a buggy game+driver combo.

**Comment by FeralChild64** [comment]
I don’t remember right now - did you set `machine = svga_s3trio64` in DOSBox-X?

**Comment by interloper98** [comment]
> I don’t remember right now - did you set `machine = svga_s3trio64` in DOSBox-X?

I think that's right. I used the default `machine = svga_s3`, which is equal to `svga_s3trio64` in DOSBox-X. 

When I checked that, turns out DOSBox-X does support `machine = svga_s3trio64v+`, and setting that produces the correct results:

![2024-Jun-16_22-51-15](https://github.com/dosbox-staging/dosbox-staging/assets/165201780/3b11b7c2-7e98-4862-af93-ef5569f167f9)

Thanks for raising this point @FeralChild64, as it solves this dillema. This makes sense, because the 1.70.04 driver is specifically for the Trio64-V+ and Trio64-V2 cards, which is a match for this DOSBox-X machine `svga_s3trio64v+`. 

But in the case of Staging's `svga_s3` aka Trio64 (non-V), I wasn't able to find a 1.70.04 driver for the non-V chip types (1.70.04's are for the V+ and/or V2) -- so it makes sense that we've got a mismatch between hardware and driver.

So it seems if we want to support the 1.70.04 driver, we need to upgrade the emu to TrioV+/V2 features and import those from DOSBox-X?

**Comment by johnnovak** [comment]
> Thanks for raising this point @FeralChild64, as it solves this dillema. This makes sense, because the 1.70.04 driver is specifically for the Trio64-V+ and Trio64-V2 cards, which is a match for `svga_s3trio64v+`.
> 
> But in the case of Staging's `svga_s3` aka Trio64 (non-V), I wasn't able to find a 1.70.04 driver for the non-V chip types (1.70.04's are for the V+ and/or V2) -- so it makes sense that we've got a mismatch between hardware and driver.

That all makes sense. Thanks for taking a detailed look into this @interloper98 and getting to the bottom of it 🎉 

> I guess we need to upgrade the emu to Trio64 **V+/V2** features and import those from DOSBox-X If we want to support the 1.70.04 driver?

Okay, so let's take a step back here before we determine we "need" to do anything 😎 

What we need to have some answers to the following questions first to understand the benefits and the costs (!) of adding V+/V2 support:

- **For DOS:** What benefits would the Trio64V+ emulation bring to the table for DOS games? As far as I know, the benefit is exactly zero.

- **For Windows 3.x:** What does the Trio64V+ emulation bring to the table for Win 3.x games? We know of this single game that has issues with the the later driver, but works perfectly fine with our current Trio64 emulation plus the v1.41B5 driver. @Grounded0 mentioned the H/W acceleration on the Trio64V+ yields some performance benefits—in the synthetic benchmarks. However, does any game use the XGA features? Does any game actually run significantly faster on the Trio64V+, making this exercise worthwhile? We're not in the business of making benchmarks show higher numbers 😎 

- Most importantly, is there any game that _will not run_ without Trio64V+ support at all? I'd bet good money on it the answer is a no.

- How complex is adding Trio64V+ support? If it's not too complicated, and it's separated from the existing Trio64 code, it can be considered. However, if the two are intertwined and/or the V+/V2 patch would be big, we would need some _very good reasons_ for adding support for it. The VGA code is complex enough without adding anything to it; we need very good reasons to complicate it any further.

So, I'm not against it, but we need justification. And that some Win 3.x game will now run the same way on 20k cycles than 30k cycles before without the XGA patch is not a good justification, especially if the code is non-trivial.

**Comment by Grounded0** [comment]
@johnnovak 

Any game that uses WinG utilises XGA features:

https://en.wikipedia.org/wiki/WinG

**Comment by johnnovak** [comment]
> @johnnovak
> 
> Any game that uses WinG utilises XGA features:
> 
> https://en.wikipedia.org/wiki/WinG

Thanks. If that list is complete, it's a slim list. Again, do using the XGA features have tangible benefits? If the only benefit is some drawing operations are a bit faster but the game runs fast enough without it anyway, I think we should not bother with it.

**Comment by Grounded0** [comment]
Well like i said XGA currently works just fine as SFPD Homicide is unplayable with WinG enabled.

**Comment by johnnovak** [comment]
> SFPD Homicide currently turns close to unplayable as the game progresses unless WinG is enabled.

But does it run fine with WinG enabled on the v1.41B5 drivers? Do we know v1.41B5 is in any way problematic with WinG?

> Well like i said XGA currently works just fine as SFPD Homicide is unplayable with WinG enabled.

I'm confused, you said a different thing above: "unless WinG enabled". Which implies to me, just enable WinG on the 1.41B5 driver and it's fine?

I'm not trying to be difficult, just want to understand if we really need this or not. Then we can make an informed decision.

We don't seem to have a ticket for SFPD Homicide, or at least I can't find it. Could you raise one for it then?

**Comment by Grounded0** [comment]
> But does it run fine with WinG enabled on the v1.41B5 drivers? Do we know v1.41B5 is in any way problematic with WinG?

Yes it does. No, i haven't faced any incompatibility problems yet by using WinG with v1.41B5 drivers.

**Comment by Grounded0** [comment]
As for SFPD Homicide it is not our bug but the game just needs WinG enabled for it to remain playable as lists grow as the game progresses and that's it.

I mentioned the fix here:

https://github.com/dosbox-staging/dosbox-staging/wiki/Game-issues#sfpd-homicide

As for this ticket since 1.70 driver is indicated for S3 Trio64V+/V2 then its not our fault it doesn't work properly since we're only emulating S3 Trio64 and this ticket could be closed.

**Comment by joncampbell123** [comment]
I'm using the S3 864 drivers from the Vogons driver library and I know of one case where I'm not certain whether the driver or DOSBox-X is at fault here. Look at the bottom half of the screen.

https://github.com/dosbox-staging/dosbox-staging/assets/6245486/c58852e6-38ab-4628-8ed0-3c3981e6a741

![setup_000](https://github.com/dosbox-staging/dosbox-staging/assets/6245486/a656b319-71c2-43ae-ac76-732d4b3b524d)

Notice that the same configuration but with Window 95 doesn't have the same rendering issue.

**Comment by joncampbell123** [comment]
Eh, setting machine=svga_s3vision864 in DOSBox-X makes some of the errors go away...

**Comment by interloper98** [comment]
> The demo definitely used to work correctly on my home PC in mid 90’s, equipped with Phoenix S3 Trio 64 <snip> v1.70.04 for S3 Trio 32/64 can be found here https://www.vogons.org/viewtopic.php?f=9&t=95939 (2nd post). It’s a different driver than v1.70.04 for S3 Trio64V.

@FeralChild64, that's a great find - I didn't see that driver in Vogon's official driver library (http://vogonsdrivers.com/index.php?catid=26); they should add it.  I suppose it's S3's final driver for the S3 Trio 32/64; I'm not seeing any notes or a changelog in the zip, but it would be great to see which fixes they made versus the v1.41B5.

> Look at the bottom half of the screen

@joncampbell123, this is really helpful - I will poke around a bit more, and double check against Wheel of Fortune, too. Hey, if this were a Star Trek game, Vana's just being beamed up :-)

From @Grounded0 and @johnnovak's discussion, we'll try for just having full featured S3 Trio 32/64, but not go beyond that.

**Comment by FeralChild64** [comment]
Tried the Critical Mass demo once again on:
- DOSBox X (current GIT), machine type s3trio64
- DOSBox Staging (current GIT)
- 86Box v4.1.1, emulating Phoenix S3 Trio 64
Guest resolution: 800x600, 64k colors, small fonts.

Results:
- drivers v1.42B22 (from www.dosdays.co.uk) - works, with every emulator (I couldn’t install the drivers on DOSBox Staging due to installer expecting Z:\CONFIG.SYS file - so I have copied the WINDOWS directory with installed drivers from within 86Box disk image)
- drivers v1.70.04 - game works correctly only with 86Box; both DOSBox X and DOSBox Staging exhibit the issue described in this bug report (missing ships)

**Comment by FeralChild64** [comment]
One more observation: switching DOSBox X to machine `svga_s3trio64v+` fixes the problem with Critical Mass game demo and Trio 32/64 driver v1.70.04.

**Comment by johnnovak** [comment]
Hmm, for all we know 86box might have a bug that negates the driver bug 😆 It's all a bit inconclusive.

**Comment by Grounded0** [comment]
S3 Trio64 driver release v1.70.04:

[w3117004.zip](https://github.com/user-attachments/files/15878266/w3117004.zip)

Safe now if Vogons goes down.

**Comment by Grounded0** [comment]
My suggestion is to test on hardware and see what the correct behaviour should be. That could be however impacted by video BIOS ROM release on the card. 🤷‍♂️

**Comment by johnnovak** [comment]
> My suggestion is to test on hardware and see what the correct presentation should be. That could be however impacted by the video BIOS ROM release on the card. 🤷‍♂️

Yeah, that's what I'm saying—is it worth it? Reality check: we're talking about one bottom-of-the-barrel Win3.x shareware game... that works fine with our known-good S3 drivers.

That's not enough motivation for me to take apart my MMX box, change VGA cards, test things over and over, etc. For the record, I do have these S3 cards. I think I'm using the S3 Virge/DX in my MMX box currently.

- S3 Trio32 (PC Partner, 86C732-P)
- S3 Trio64 V+ (?, 86C765)
- S3 Virge/DX (Diamond Stealth 3D 2000 Pro, 86C375, 2MB)
- S3 Trio64 V2/DX (Elsa, 86C775, 1MB)
- S3 Trio64 V+ (Elsa, 86C765, 2MB)

Also, what even does the newer driver version bring to the table? Usually, they only release these to add support for new chipsets, etc., so I don't assume newer driver == better for our use case.

So, my suggestion is to leave this open, just in case, and it's good to have it documented, but IMO it requires no action unless there's a significant number of non-bottom-of-the-barrel games that absolutely need proper XGA support.

**Comment by Grounded0** [comment]
Press release
S3 Unveils its Cooperative Accelerator Architecture:

https://web.archive.org/web/19970607065045/http://www.s3.com/corporate/news/new_products/061395.html

"The Trio64V+ is also the first multimedia accelerator to utilize the complete Windows 95 DirectDraw feature set for 2D graphics and video."

Datasheet
S3 Trio64V+ Integrated Graphics/Video Accelerator:

http://www.bitsavers.org/components/s3/DB018-A_Trio64V+_Integrated_Graphics_Video_Accelerator_Jul95.pdf

VGA Museum
S3 Trio64V+:

https://www.vgamuseum.info/index.php/component/k2/item/359-s3-trio64v

**Comment by Grounded0** [comment]
One solution would be to have `svga_s3` dialing in the new identificator for Windows 3.x since the driver doesn't utilise all on-die capabilities the Windows 9x driver expects and have `svga_s3_old` with the old identificator for Windows 9x.

**Comment by emxd** [comment]
The v1.41B5 drivers seem to have no hardware cursors.

https://github.com/user-attachments/assets/a5ad3540-44a5-4b65-b2ad-78a5075546e6

v1.41B5 - cursor flickering
v1.42B22 - cursor flickering
v1.70.04 - cursor works ok
